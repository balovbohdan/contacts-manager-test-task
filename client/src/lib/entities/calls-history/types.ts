import {CallType} from './CallType';

export type CallsHistory = {[type:string]:Item};

export type Item = {
    id:number;
    time:number;
    type:CallType;
    contactId:number;
};