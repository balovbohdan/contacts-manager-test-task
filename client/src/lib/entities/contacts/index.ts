import * as utils from './utils';
export {findContact} from './find-contact';

export {utils};