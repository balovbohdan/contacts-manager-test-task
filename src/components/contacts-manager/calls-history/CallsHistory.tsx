import * as React from 'react';

import {ModalWindow} from '@components/modal-window';

import css from './CallsHistory.css';

export const CallsHistory = () =>
    <ModalWindow title='Title' close={() => alert('close')}>
        <div className={css.main}>
            <div className={css.body}>
            </div>
        </div>
    </ModalWindow>;