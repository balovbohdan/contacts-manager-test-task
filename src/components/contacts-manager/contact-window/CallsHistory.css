.main {
    overflow: hidden;
    max-height: 200px;
    margin: 20px 0 10px;
}

.scrollWrapper {
    overflow-y: auto;
    max-height: 200px;
}

.item {
    width: 100%;
    height: 40px;

    display: flex;
    align-items: center;
    justify-content: flex-start;

    cursor: pointer;
    margin-top: 10px;
    overflow: hidden;
    transition: var(--trans);
    background: var(--color-blue);
    box-shadow: 0 1px 2px rgba(0, 0, 0, .2), 0 3px 3px rgba(0, 0, 0, .2);

    font-size: .8rem;
}

.item:hover {
    background-color: rgba(22, 50, 79, .8);
}

.itemIcon {
    width: 40px;
    height: 100%;

    display: flex;
    align-items: center;
    justify-content: center;

    background-color: rgba(255, 255, 255, .1);
}

.itemIcon_img {
    width: calc(100% - 10px);
    height: calc(100% - 10px);
}

.item_toName {
    margin-left: 10px;
}

.callType {
    width: 15px;
    height: 15px;

    margin-left: 10px;

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
}

.callType__income {
    background-image: url(~@img/income-call.svg);
}

.callType__outcome {
    background-image: url(~@img/outcome-call.svg);
}