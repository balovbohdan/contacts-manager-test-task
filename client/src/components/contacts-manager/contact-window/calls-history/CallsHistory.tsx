import * as React from 'react';
import classnames from 'classnames';

import {CallType, T} from '@lib/entities/calls-history';
import {findContact} from '@lib/entities/contacts/utils';
import {Contact, Contacts} from '@lib/entities/contacts/contact/types';

import {FetchCallsHistory} from '@storehouse/actions/contacts/types';

import * as utils from '../utils';
import * as css from './CallsHistory.css';
import {DataProvider} from './DataProvider';

type Props = {
    contactId:number;
    contacts:Contacts;
    history:T.CallsHistory;

    fetchCallsHistory:FetchCallsHistory;
};

type ItemsProps = {
    contacts:Contacts;
    history:T.CallsHistory;
};

type ItemProps = {
    time:number;
    type:CallType;
    toContact:Contact;
};

export type CallTypeIconProps = {
    type:CallType;
};

export const CallsHistory = (props:Props) =>
    <DataProvider
        contactId={props.contactId}
        fetchCallsHistory={props.fetchCallsHistory}>
        <div className={css.main}>
            <div className={css.scrollWrapper}>
                <Items history={props.history} contacts={props.contacts}/>
            </div>
        </div>
    </DataProvider>;

const Items = ({history, contacts}:ItemsProps) => {
    const mapper = ({id, time, type, contactId}) => {
        try {
            const contact = findContact(contactId, contacts);

            return <Item key={id} time={time} type={type} toContact={contact}/>;
        } catch (e) {
            return null;
        }
    };

    const items = Object.values(history)
        .reverse()
        .map(mapper)
        .filter(item => !!item);

    return <>{items}</>;
};

const Item = ({time, type, toContact:{img, name, phone}}:ItemProps) => {
    const timePrepared = utils.prepareTime(time);
    const title = `${phone}; ${timePrepared}`;

    const iconStyle = {
        backgroundImage: `url(${img})`
    };

    return (
        <div title={title} className={css.item}>
            <div className={css.itemIcon}>
                <div style={iconStyle} className={css.itemIcon_img}/>
            </div>
            <CallTypeIcon type={type}/>
            <p className={css.item_toName}>{timePrepared}</p>
        </div>
    );
};

const CallTypeIcon = ({type}:CallTypeIconProps) => {
    const className = type === CallType.INCOME
        ? css.callType__income
        : css.callType__outcome;

    const classNames = classnames([
        className,
        css.callType
    ]);

    return <div className={classNames}/>;
};