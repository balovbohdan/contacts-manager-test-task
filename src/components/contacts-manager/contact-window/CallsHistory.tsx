import * as React from 'react';
import classnames from 'classnames';

import {CallType, T} from '@lib/entities/calls-history';
import {findContact} from '@lib/entities/contacts/utils';
import {Contact, Contacts} from '@lib/entities/contacts/contact/types';

import * as utils from './utils';
import css from './CallsHistory.css';

type Props = {
    contacts:Contacts;
    history:T.CallsHistory;
};

type ItemsProps = {
    contacts:Contacts;
    history:T.CallsHistory;
};

type ItemProps = {
    time:number;
    type:CallType;
    toContact:Contact;
};

export type CallTypeIconProps = {
    type:CallType;
};

export const CallsHistory = ({history, contacts}:Props) =>
    <div className={css.main}>
        <div className={css.scrollWrapper}>
            <Items history={history} contacts={contacts}/>
        </div>
    </div>;

const Items = ({history, contacts}:ItemsProps) => {
    const mapper = ({id, time, type, contactId}) => {
        try {
            const contact = findContact(contactId, contacts);

            return <Item key={id} time={time} type={type} toContact={contact}/>;
        } catch (e) {
            return null;
        }
    };

    const items = Object.values(history)
        .reverse()
        .map(mapper)
        .filter(item => !!item);

    return <>{items}</>;
};

const Item = ({time, type, toContact:{img, name, phone}}:ItemProps) => {
    const timePrepared = utils.prepareTime(time);
    const title = `${phone}; ${timePrepared}`;

    const iconStyle = {
        backgroundImage: `url(${img})`
    };

    return (
        <div title={title} className={css.item}>
            <div className={css.itemIcon}>
                <div style={iconStyle} className={css.itemIcon_img}/>
            </div>
            <CallTypeIcon type={type}/>
            <p className={css.item_toName}>{timePrepared}</p>
        </div>
    );
};

const CallTypeIcon = ({type}:CallTypeIconProps) => {
    const className = type === CallType.INCOME
        ? css.callType__income
        : css.callType__outcome;

    const classNames = classnames([
        className,
        css.callType
    ]);

    return <div className={classNames}/>;
};