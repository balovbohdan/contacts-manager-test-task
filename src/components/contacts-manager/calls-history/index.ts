import {CallsHistory} from './CallsHistory';

export {CallsHistory};
export default CallsHistory;