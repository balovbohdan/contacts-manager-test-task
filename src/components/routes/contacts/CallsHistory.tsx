import * as React from 'react';

import {CallsHistory as Base} from '@components/contacts-manager';

type Props = {
    active:boolean;
};

export const CallsHistory = ({active}:Props) =>
    !active
        ? null
        : <Base/>;