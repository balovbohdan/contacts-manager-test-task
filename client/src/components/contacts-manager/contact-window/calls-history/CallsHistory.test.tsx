import * as React from 'react';
import {mount, shallow} from 'enzyme';

import {CallsHistory} from './CallsHistory';
import {CallType} from '@lib/entities/calls-history';

describe('CallsHistory', () => {
    const propsDef = {
        contactId: 1,
        fetchCallsHistory: jest.fn(),
        contacts: [{
            id: 1,
            name: 'name',
            img: '/user.svg',
            phone: '+380961514231'
        }],
        history: {
            1: {
                id: 1,
                contactId: 1,
                time: 1568728308621,
                type: CallType.INCOME
            }
        }
    };

    it('Should match snapshot', () => {
        const props = { ...propsDef };
        const history = shallow(<CallsHistory {...props}/>);

        expect(history).toMatchSnapshot();
    });

    it('Should call history fetcher function', () => {
        const props = {
            ...propsDef,
            fetchCallsHistory: jest.fn()
        };

        mount(<CallsHistory {...props}/>);

        expect(props.fetchCallsHistory).toBeCalledTimes(1);
    });
});