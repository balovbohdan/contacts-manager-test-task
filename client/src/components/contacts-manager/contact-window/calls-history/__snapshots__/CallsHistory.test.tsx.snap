// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallsHistory Should match snapshot 1`] = `
<Component
  contactId={1}
  fetchCallsHistory={[MockFunction]}
>
  <div
    className="main"
  >
    <div
      className="scrollWrapper"
    >
      <Items
        contacts={
          Array [
            Object {
              "id": 1,
              "img": "/user.svg",
              "name": "name",
              "phone": "+380961514231",
            },
          ]
        }
        history={
          Object {
            "1": Object {
              "contactId": 1,
              "id": 1,
              "time": 1568728308621,
              "type": 1,
            },
          }
        }
      />
    </div>
  </div>
</Component>
`;
